tests/fuzz/s2n_deserialize_resumption_state_test.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* Target Functions: s2n_deserialize_resumption_state
+ * This test fuzzes the deserialization logic for our session tickets. This occurs after
+ * the ticket has been decrypted with the server ticket key. Technically it's not
+ * necessary to fuzz decrypted values, however, in the event that an attacker is 
+ * able to get a valid ticket key, we want to make sure these functions can't cause damage.
+ */
+
+#include "api/s2n.h"
+#include "s2n_test.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_resume.h"
+#include "utils/s2n_safety.h"
+
+int s2n_fuzz_test(const uint8_t *buf, size_t len)
+{
+    /* We need at least one byte of input to set parameters */
+    S2N_FUZZ_ENSURE_MIN_LEN(len, 1);
+
+    DEFER_CLEANUP(struct s2n_stuffer fuzzed_ticket = { 0 }, s2n_stuffer_free);
+    POSIX_GUARD(s2n_stuffer_alloc(&fuzzed_ticket, len));
+    POSIX_GUARD(s2n_stuffer_write_bytes(&fuzzed_ticket, buf, len));
+
+    /* Pull a byte off the libfuzzer input and use it to set parameters */
+    uint8_t randval = 0;
+    POSIX_GUARD(s2n_stuffer_read_uint8(&fuzzed_ticket, &randval));
+    POSIX_GUARD(s2n_stuffer_reread(&fuzzed_ticket));
+    POSIX_GUARD(s2n_stuffer_rewrite(&fuzzed_ticket));
+
+    /* There are only a few valid formats for session tickets; this ensures the
+     * format version is at or below S2N_SERIALIZED_FORMAT_TLS12_V3, which will
+     * keep the test checking mostly valid paths. */
+    randval = randval % S2N_SERIALIZED_FORMAT_TLS12_V3;
+    POSIX_GUARD(s2n_stuffer_write_uint8(&fuzzed_ticket, randval));
+    /* We have to put the write cursor back */
+    POSIX_GUARD(s2n_stuffer_skip_write(&fuzzed_ticket, len - 1));
+
+    /* A session ticket is sent along with the Client Hello, so there's not much set up needed for the server */
+    DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+    POSIX_ENSURE_REF(server_conn);
+    POSIX_GUARD(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+
+    uint8_t test_data[] = "test psk identity";
+    struct s2n_blob test_blob   = { .data = test_data, .size = sizeof(test_data) };
+
+    /* Ignore the result of this function */
+    s2n_result_ignore(s2n_deserialize_resumption_state(server_conn, &test_blob, &fuzzed_ticket));
+
```

Comment:
So we're just verifying that no inputs can crash the process?

Reply:
Yes, given fuzzed input the function should fail and not crash. There *might* be an input that succeeds, so we don't have an EXPECT_FAILURE macro around this.

Reply:
It's fuzzing the plaintext ticket. So it would only succeed if it created a valid ticket, which would very very rare.

Reply:
If there's no authentication on it then it wouldn't be rare, especially for a gray-box fuzzer like libfuzzer. But if we're fuzzing the post-decrypt plaintext then we likely won't be able to have any additional oracles beyond "did it crash".

Reply:
Hmmm yeah actually I would be interested in how frequently libfuzzer could create a valid ticket. Just in general, we need better insights into what the fuzzer is doing.

Reply:
My guess is it would create a valid ticket _very_ quickly. That's the huge advantage of gray-box fuzzing. See this article as an example: https://lcamtuf.blogspot.com/2014/11/pulling-jpegs-out-of-thin-air.html?m=1